### PR TITLE
[WIP] fix(render): Fix for #1197

### PR DESCRIPTION
**Pull request by majo44**
Fix for #1197 but this depends on changes for https://github.com/developit/preact/issues/654

### CODE REVIEW

File: src/with-component.js [hunk 1]
```diff
@@ -1,9 +1,19 @@
...
 import { withUnique } from './with-unique';
+import { isDefined } from './define';
+
+let oldVnodeHook = options.vnode;
+options.vnode = (vnode) => {
+  if (oldVnodeHook) oldVnodeHook(vnode);
+  vnode.skipChildren =
+    typeof vnode.nodeName === 'string' &&
+    vnode.nodeName.indexOf('-') > 0 &&
+    isDefined(vnode.nodeName);
```
Comment (treshugart) [inline src/with-component.js]: Why not just use `customElements.get(vnode.nodeName)`?

**Comment by treshugart** [comment]
While I can see a need to do something like this, I don't think it belongs in core Skate as the core focuses on the happy path of Preact rendering to a shadow root. If you want to have a custom `renderRoot`, then it should be up to the implementation (or non-core extension) to do what is being done here.

Thoughts?

**Comment by majo44** [comment]
@treshugart
 Sorry I do not understand this point of view.
Library is exposing some API (Component, renderCallback, renderRooot,....)
I used this core API and figure out that this API is not work proper.
I even created the test case which is exactly shows that the core API is not working as expected.
So there are two possibilities: 
1. API is not designed well, (and eg renderRoot should be removed) or 
2. this is just a bug, which have to be fixed (if is not fixable or you do not want to fix it please look point 1.).

**Comment by treshugart** [comment]
The core of the problem here is that Preact will blow away anything it doesn't know about. This is a side-effect of the diff / patch algorithm of most virtual DOM libraries and is not a bug. I realise you probably already know this, but I want to be clear for the sake of posterity.

Skate provides `renderRoot` as an escape hatch for you to render to whatever you want but does not make any assumptions as to what that is because it could be anything. Skate is not misbehaving here; this design is intentional. The API line is drawn here.

Modifying Preact `options` here can also have side-effects if the same Preact instance is shared in other places. We had problems when doing this with Incremental DOM and the workarounds were hard to understand and difficult to maintain.

I like the idea of documenting how to do this and even going as far to provide a non-core way of doing it as a separate implementation but I don't want to maintain it as part of the core. The focus of this part of the API is to try and coerce the consumer to use Shadow DOM because this is the ideal path, but give them an escape hatch to do whatever they want, at their own risk because sometimes that is necessary. This is where the API line is drawn.

If this makes the API "not designed well," then so be it.

**Comment by majo44** [comment]
@treshugart 
" This is a side-effect of the diff / patch algorithm of most virtual DOM libraries and is not a bug. "
It is not full true. Incrementl-dom have a skip/skipNode which address this problem. There are also https://github.com/Matt-Esch/virtual-dom/blob/master/docs/thunk.md, https://github.com/snabbdom/snabbdom#thunks. In other VDOMs the live cycle interceptors can be used for the same purpose eg: https://github.com/patrick-steele-idem/morphdom#morphdomfromnode-tonode-options--node. Even Preact for own Components supports this by the shouldComponentUpdate, but is to hi level in this case. 

I agree that approach like in this PR it is not perfect. And I just realized that such solution in case of mix shadow & light DOM components on one page will cause a next bug (the slots will be not working in ShadowDOM, because light DOM will be not updated so shadow DOM will not update the slots).

When I said that Api is not designed well I had in mind that maybe we should consider different approach.

So lets think forward. I will look for other options.

**Pull request closed without merging**